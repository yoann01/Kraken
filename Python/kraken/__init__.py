"""Kraken Framework."""

import logging
import os

__all__ = ['core', 'helpers', 'plugins', 'ui']


krakenPath = os.path.abspath(os.path.join(os.path.dirname(os.path.realpath(__file__)), '..', '..'))
if os.environ.get('KRAKEN_PATH', None) is None:
    os.environ['KRAKEN_PATH'] = krakenPath

krakenExtsPath = os.path.join(krakenPath, 'Exts')
if 'FABRIC_EXTS_PATH' in os.environ:
    if krakenExtsPath not in os.environ['FABRIC_EXTS_PATH']:
        os.environ['FABRIC_EXTS_PATH'] = krakenExtsPath + os.pathsep + os.environ['FABRIC_EXTS_PATH']

krakenPresetsPath = os.path.join(krakenPath, 'Presets', 'DFG')
if 'FABRIC_DFG_PATH' in os.environ:
    if krakenPresetsPath not in os.environ['FABRIC_DFG_PATH']:
        os.environ['FABRIC_DFG_PATH'] = krakenPresetsPath + os.pathsep + os.environ['FABRIC_DFG_PATH']


logging.basicConfig(format='[KRAKEN] %(levelname)s: %(message)s', level=logging.INFO)

# Custom inform level for use with UI label getting added to the status bar.
logging.INFORM = 25
logging.addLevelName(logging.INFORM, 'INFORM')
logging.Logger.inform = lambda inst, msg, *args, **kwargs: inst.log(logging.INFORM, msg, *args, **kwargs)


def release():
    print "RELEASE THE KRAKEN!!!!"
