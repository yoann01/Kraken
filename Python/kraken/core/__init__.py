"""Kraken Core."""

VERSION_MAJOR = 1
VERSION_MINOR = 2
VERSION_BUILD = 6
VERSION_SUFFIX = ""


def getVersion():
    """Contatenates the version globals and returns the current version of
    Kraken.

    Returns:
        str: Current version of Kraken.

    """

    versionString = str(VERSION_MAJOR) + "." + str(VERSION_MINOR) + "." + str(VERSION_BUILD)
    if VERSION_SUFFIX:
        versionString = versionString + "-" + VERSION_SUFFIX

    return versionString
