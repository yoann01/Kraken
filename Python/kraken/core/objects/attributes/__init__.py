"""Kraken Attributes."""

__all__ = ['attribute_group',
           'attribute',
           'bool_attribute',
           'color_attribute',
           'scalar_attribute',
           'integer_attribute',
           'string_attribute']
