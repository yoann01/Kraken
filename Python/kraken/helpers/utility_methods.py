from kraken.core.maths.math_object import MathObject
from kraken.core.maths.vec3 import Vec3
from kraken.core.maths.xfo import Xfo
from kraken.core.maths.quat import Quat

from kraken.core.maths import decodeValue


def logHierarchy(kObject):
    """Traverses the given Kraken hierarchy and logs the names of all the objects.

    Args:
        Object: The object to start logging on.

    """

    print kObject.getDecoratedPath()
    for i in xrange(kObject.getNumChildren()):
        child = kObject.getChildByIndex(i)
        logHierarchy(child)


def __convertFromJSON(jsonData):

    if type(jsonData) is list:
        newList = []
        for item in jsonData:
            newList.append(__convertFromJSON(item))
        return newList
    elif type(jsonData) is dict:
        if '__mathObjectClass__' in jsonData.keys():
            return decodeValue(jsonData)
        for key, value in jsonData.iteritems():
            jsonData[key] = __convertFromJSON(value)
    return jsonData


def prepareToLoad(jsonData):
    """Prepares the json data for loading into kraken.

    Args:
        jsonData (dict): The JSON data to be prepared.

    Returns:
        dict: The prepared JSON hierarchy.

    """

    return __convertFromJSON(jsonData)


def __convertToJSON(jsonData):

    if isinstance(jsonData, MathObject):
        return jsonData.jsonEncode()
    elif type(jsonData) is list:
        newList = []
        for item in jsonData:
            newList.append(__convertToJSON(item))
        return newList
    elif type(jsonData) is dict:
        newDict = {}
        for key, value in jsonData.iteritems():
            newDict[key] = __convertToJSON(value)
        return newDict
    return jsonData


def prepareToSave(jsonData):
    """Prepares the json data for serialization.

    Args:
        jsonData (dict): The JSON data to be prepared.

    Returns:
        dict: The prepared JSON hierarchy.

    """

    return __convertToJSON(jsonData)


def __mirrorData(jsonData, plane):

    if isinstance(jsonData, Vec3):
        newVec3 = Vec3(jsonData)
        if plane == 0:
            newVec3.x = -newVec3.x
        elif plane == 1:
            newVec3.y = -newVec3.y
        elif plane == 2:
            newVec3.z = -newVec3.z
        return newVec3

    if isinstance(jsonData, Quat):
        newQuat = Quat(jsonData)
        newQuat.mirror(plane)
        return newQuat

    elif isinstance(jsonData, Xfo):
        newXfo = Xfo(jsonData)
        if plane == 0:
            newXfo.tr.x = -newXfo.tr.x
        elif plane == 1:
            newXfo.tr.y = -newXfo.tr.y
        elif plane == 2:
            newXfo.tr.z = -newXfo.tr.z

        newXfo.ori.mirror(plane)
        return newXfo

    elif type(jsonData) is list:
        newList = []
        for item in jsonData:
            newList.append(__mirrorData(item, plane))
        return newList

    elif type(jsonData) is dict:
        newDict = {}
        for key, value in jsonData.iteritems():
            newDict[key] = __mirrorData(value, plane)
        return newDict

    return jsonData


def mirrorData(jsonData, plane):
    """Prepares the json data for serialization.

    Args:
        jsonData (dict): The JSON data to be prepared.
        plane (int): The plane to mirror across.

    Returns:
        dict: The prepared JSON hierarchy.

    """

    return __mirrorData(jsonData, plane)
