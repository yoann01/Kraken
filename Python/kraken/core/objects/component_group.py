"""Kraken - objects.component_group module.

Classes:
ComponentGroup -- Hiearchy group representation.

"""

from kraken.core.objects.object_3d import Object3D


class ComponentGroup(Object3D):
    """ComponentGroup object."""

    def __init__(self, name, component, parent=None, metaData=None):
        super(ComponentGroup, self).__init__(name, parent=parent, metaData=metaData)

        self.setComponent(component)

        self.setShapeVisibility(False)
        self.lockRotation(True, True, True)
        self.lockScale(True, True, True)
        self.lockTranslation(True, True, True)


    # =============
    # Name Methods
    # =============
    def getName(self):
        """Gets the decorated name of the object.

        Returns:
            str: Decorated name of the object.

        """

        # During construction of the base class, the name is tested before
        # the component is assigned.
        if self.getComponent() is not None:
            # The ComponentGroup's name should always match the component's name.
            return self.getComponent().getName()
        else:
            return super(ComponentGroup, self).getName()


    def getNameDecoration(self):
        """Gets the decorated name of the object.

        Returns:
            str: Decorated name of the object.

        """

        if self.getComponent() is not None:
            # The ComponentGroup's name should always match the component's name.
            return self.getComponent().getNameDecoration()
        else:
            return super(ComponentGroup, self).getNameDecoration()
