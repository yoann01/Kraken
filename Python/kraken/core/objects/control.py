"""Kraken - objects.Controls.Control module.

Classes:
Control - Base Control.

"""

from kraken.core.configs.config import Config
from kraken.core.maths.constants import AXIS_NAME_TO_TUPLE_MAP
from kraken.core.maths import Euler, Quat, Vec3, Xfo
from kraken.core.maths import Math_degToRad
from kraken.core.objects.curve import Curve
from kraken.core.objects.ctrlSpace import CtrlSpace


class Control(Curve):
    """Base Control object."""

    def __init__(self, name, parent=None, shape="null", scale=1.0, flags=None, metaData=None):
        """Initializes control object.

        Args:
            name (str): Name of the constraint.
            parent (Object): Parent object of this object.

        """

        super(Control, self).__init__(name, parent=parent, flags=flags, metaData=metaData)
        self.shape = shape

        self.setShape(shape)

        self.scalePoints(Vec3(scale,scale,scale))

    # ==============
    # Shape Methods
    # ==============
    def getShape(self):
        """Returns the shape that the control was set to.

        Returns:
            str: Name of the shape that was set.

        """

        return self.shape

    def setShape(self, shape):
        """Sets the shape of the control to the one specified.

        Args:
            shape (str): the desired shape of the control.

        Returns:
            bool: True if successful.

        """

        config = Config.getInstance()
        configShapes = config.getControlShapes()
        if shape not in configShapes.keys():
            raise KeyError("'" + shape + "' is not a valid shape in the loaded config of class [" + config.__class__.__name__ + "]")

        self.shape = shape
        self.setCurveData(configShapes[shape])

        return True


    # ==============
    # Align Methods
    # ==============
    def alignOnXAxis(self, negative=False):
        """Aligns the control shape on the X axis.

        Args:
            negative (bool): Whether to align the control on the negative X axis.

        Returns:
            bool: True if successful.

        """

        furthest = 0.0

        curveData = list(self.getCurveData())

        for eachSubCurve in curveData:
            for eachPoint in eachSubCurve["points"]:

                if negative is False:

                    if eachPoint[0] < furthest:
                        furthest = eachPoint[0]

                else:

                    if eachPoint[0] > furthest:
                        furthest = eachPoint[0]


        offset = 0.0 - furthest

        for eachSubCurve in curveData:
            for eachPoint in eachSubCurve["points"]:
                eachPoint[0] += offset

        self.setCurveData(curveData)

        return True

    def alignOnYAxis(self, negative=False):
        """Aligns the control shape on the Y axis.

        Args:
            negative (bool): Whether to align the control on the negative Y axis.

        Returns:
            bool: True if successful.

        """

        furthest = 0.0

        curveData = list(self.getCurveData())

        for eachSubCurve in curveData:
            for eachPoint in eachSubCurve["points"]:

                if negative is False:

                    if eachPoint[1] < furthest:
                        furthest = eachPoint[1]

                else:

                    if eachPoint[1] > furthest:
                        furthest = eachPoint[1]

        offset = 0.0 - furthest

        for eachSubCurve in curveData:
            for eachPoint in eachSubCurve["points"]:
                eachPoint[1] += offset

        self.setCurveData(curveData)

        return True

    def alignOnZAxis(self, negative=False):
        """Aligns the control shape on the Z axis.

        Args:
            negative (bool): Whether to align the control on the negative Z axis.

        Returns:
            bool: True if successful.

        """

        furthest = 0.0

        curveData = list(self.getCurveData())

        for eachSubCurve in curveData:
            for eachPoint in eachSubCurve["points"]:

                if negative is False:

                    if eachPoint[2] < furthest:
                        furthest = eachPoint[2]

                else:

                    if eachPoint[2] > furthest:
                        furthest = eachPoint[2]

        offset = 0.0 - furthest

        for eachSubCurve in curveData:
            for eachPoint in eachSubCurve["points"]:
                eachPoint[2] += offset

        self.setCurveData(curveData)

        return True


    # ==============
    # Scale Methods
    # ==============
    def scalePointsOnAxis(self, scale, scaleAxis="POSX"):
        """Scales the point positions from it's center along the given axis only.

        Args:
            scale (float): scale value to apply to the points.
            scaleAxis (str): which axes to scale and by what direction

        Returns:
            bool: True if successful.

        """

        # would be great if vec3 was iterable
        axis = AXIS_NAME_TO_TUPLE_MAP.get(scaleAxis)

        if axis is None:
            raise KeyError("'" + scaleAxis + "' is not a valid axis. Valid axes are: " + ','.join(AXIS_NAME_TO_TUPLE_MAP.keys()))

        scaleList = [1.0, 1.0, 1.0]
        for i, x in enumerate(axis):
            if x != 0:
                scaleList[i] = scale * axis[i]

        return self.scalePoints(Vec3(scaleList[0], scaleList[1], scaleList[2]))


    # ==============
    # Scale Methods
    # ==============
    def scalePoints(self, scaleVec):
        """Scales the point positions from it's center.

        Args:
            scaleVec (Vec3): scale value to apply to the points.

        Returns:
            bool: True if successful.

        """

        curveData = list(self.getCurveData())

        for eachSubCurve in curveData:
            for eachPoint in eachSubCurve["points"]:
                eachPoint[0] *= scaleVec.x
                eachPoint[1] *= scaleVec.y
                eachPoint[2] *= scaleVec.z

        self.setCurveData(curveData)

        return True


    # ===============
    # Rotate Methods
    # ===============
    def rotatePoints(self, xRot, yRot, zRot):
        """Rotates the points by the input values.

        Args:
            xRot (float): Euler x rotate value.
            yRot (float): Euler y rotate value.
            zRot (float): Euler z rotate value.

        Returns:
            bool: True if successful.

        """

        curveData = list(self.getCurveData())

        quatRot = Quat()
        quatRot.setFromEuler(Euler(Math_degToRad(xRot), Math_degToRad(yRot),
                                   Math_degToRad(zRot)))

        for eachSubCurve in curveData:

            for eachPoint in eachSubCurve["points"]:
                pointVec = Vec3(eachPoint[0], eachPoint[1], eachPoint[2])
                rotVec = quatRot.rotateVector(pointVec)
                eachPoint[0] = rotVec.x
                eachPoint[1] = rotVec.y
                eachPoint[2] = rotVec.z

        self.setCurveData(curveData)

        return True


    # ==================
    # Translate Methods
    # ==================
    def translatePoints(self, translateVec):
        """Translates the control points.

        Args:
            translateVec (Vec3): Translation values to apply to the points.

        Returns:
            bool: True if successful.

        """

        curveData = list(self.getCurveData())

        for eachSubCurve in curveData:
            for eachPoint in eachSubCurve["points"]:
                eachPoint[0] += translateVec.x
                eachPoint[1] += translateVec.y
                eachPoint[2] += translateVec.z

        self.setCurveData(curveData)

        return True


    # ===============
    # Helper Methods
    # ===============
    def insertCtrlSpace(self, name=None):
        """Adds a CtrlSpace object above this object

        Args:
            name (string): optional name for this CtrlSpace, default is same as
                this object

        Returns:
            object: New CtrlSpace object

        """

        if name is None:
            name = self.getName()

        newCtrlSpace = CtrlSpace(name, parent=self.getParent())
        if self.getParent() is not None:
            self.getParent().removeChild(self)

        if self.getMetaDataItem("altLocation") is not None:
            newCtrlSpace.setMetaDataItem("altLocation", self.getMetaDataItem("altLocation"))

        self.setParent(newCtrlSpace)
        newCtrlSpace.addChild(self)

        newCtrlSpace.xfo = Xfo(self.xfo)

        # To ensure that names of control spaces don't clash with controls and
        # if they do, set's the control space's name back to what it was intended
        if self.getName() == name:
            newCtrlSpace.setName(name)

        return newCtrlSpace
