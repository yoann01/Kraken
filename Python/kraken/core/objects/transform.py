"""Kraken - objects.transform module.

Classes:
Transform -- Transform representation.

"""

from kraken.core.objects.object_3d import Object3D


class Transform(Object3D):
    """Transform object."""

    def __init__(self, name, parent=None, flags=None, metaData=None):
        super(Transform, self).__init__(name, parent=parent, flags=flags, metaData=metaData)
