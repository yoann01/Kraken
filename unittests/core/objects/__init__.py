
import unittest

from attributes import suite as attributeSuite
from components import suite as componentSuite
from constraints import suite as constraintSuite
from operators import suite as operatorSuite

from test_component_group import suite as componentGroupSuite
from test_container import suite as containerSuite
from test_control import suite as controlSuite
from test_ctrlSpace import suite as ctrlSpaceSuite
from test_curve import suite as curveSuite
from test_hierarchy_group import suite as hierarchyGroupSuite
from test_joint import suite as jointSuite
from test_layer import suite as layerSuite
from test_locator import suite as locatorSuite
from test_object_3d import suite as object3DSuite
from test_rig import suite as rigSuite
from test_scene_item import suite as sceneItemSuite
from test_transform import suite as transformSuite

loadAttributeSuite = attributeSuite()
loadComponentSuite = componentSuite()
loadConstraintSuite = constraintSuite()
loadOperatorSuite = operatorSuite()

loadComponentGroupSuite = componentGroupSuite()
loadContainerSuite = containerSuite()
loadControlSuite = controlSuite()
loadCtrlSpaceSuite = ctrlSpaceSuite()
loadCurveSuite = curveSuite()
loadHierarchyGroupSuite = hierarchyGroupSuite()
loadJointSuite = jointSuite()
loadLayerSuite = layerSuite()
loadLocatorSuite = locatorSuite()
loadObject3DSuite = object3DSuite()
loadRigSuite = rigSuite()
loadSceneItemSuite = sceneItemSuite()
loadTransformSuite = transformSuite()


def suite():
    suites = [
        loadAttributeSuite,
        loadComponentSuite,
        loadConstraintSuite,
        loadOperatorSuite,
        loadContainerSuite,
        loadControlSuite,
        loadCtrlSpaceSuite,
        loadCurveSuite,
        loadHierarchyGroupSuite,
        loadJointSuite,
        loadLayerSuite,
        loadLocatorSuite,
        loadObject3DSuite,
        loadRigSuite,
        loadSceneItemSuite,
        loadTransformSuite]

    return unittest.TestSuite(suites)


if __name__ == '__main__':
    runner = unittest.TextTestRunner()
    runner.run(suite())
