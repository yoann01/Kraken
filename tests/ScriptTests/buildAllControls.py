from kraken import plugins
from kraken.core.objects.transform import Transform
from kraken.core.objects.control import Control
from kraken.helpers.utility_methods import logHierarchy

# Build
builder = plugins.getBuilder()
config = builder.getConfig()
config.setExplicitNaming(True)

ctrlParent = Transform("controls")

ctrlShapes = config.getControlShapes()

i = 0
j = 0
for k, v in ctrlShapes.iteritems():
    ctrl = Control(k + '_shape', shape=k, parent=ctrlParent)

    ctrl.xfo.tr.x = i % 5 * 2
    ctrl.xfo.tr.z = j * 2

    if i % 5 == 4:
        j += 1

    i += 1


builder.build(ctrlParent)

logHierarchy(ctrlParent)
