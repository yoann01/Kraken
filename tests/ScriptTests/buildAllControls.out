Traceback (most recent call last):
  File "tests\ScriptTests\buildAllControls.py", line 2, in <module>
    from kraken.plugins.maya_plugin import builder as mayaBuilder
  File "Python\kraken\plugins\maya_plugin\builder.py", line 18, in <module>
    from kraken.plugins.maya_plugin.utils import *
  File "Python\kraken\plugins\maya_plugin\utils\__init__.py", line 1, in <module>
    import pymel.core as pm
ImportError: No module named pymel.core

